mohawk/util.py

```diff
@@ -149,45 +151,40 @@ def parse_authorization_header(auth_header):
         'Hawk id="dh37fgj492je", ts="1367076201", nonce="NPHgnG", ext="and
         welcome!", mac="CeWHy4d9kbLGhDlkyw2Nh3PJ7SDOdZDa267KH4ZaNMY="'
     """
-    attributes = {}
+    if len(auth_header) > MAX_LENGTH:
+        raise BadHeaderValue('Header exceeds maximum length of %d' % MAX_LENGTH)
 
     # Make sure we have a unicode object for consistency.
     if isinstance(auth_header, six.binary_type):
         auth_header = auth_header.decode('utf8')
 
-    parts = auth_header.split(',')
-    auth_scheme_parts = parts[0].split(' ')
-    if 'hawk' != auth_scheme_parts[0].lower():
+    scheme, attributes_full = auth_header.split(' ', 1)
+
+    if scheme.lower() != 'hawk':
         raise HawkFail("Unknown scheme '{scheme}' when parsing header"
                        .format(scheme=auth_scheme_parts[0].lower()))
 
-    # Replace 'Hawk key: value' with 'key: value'
-    # which matches the rest of parts
-    parts[0] = auth_scheme_parts[1]
 
-    for part in parts:
-        attr_parts = part.split('=')
-        key = attr_parts[0].strip()
+    attributes = {}
+
+    def replace_attribute(match):
+        '''Replace each attribute in the header with '', a correctly formatted header will then
+        by the end equal ''.
+        '''
+        key = match.group('key')
+        value = match.group('value')
         if key not in allowable_header_keys:
             raise HawkFail("Unknown Hawk key '{key}' when parsing header"
                            .format(key=key))
-
-        if len(attr_parts) > 2:
-            attr_parts[1] = '='.join(attr_parts[1:])
-
-        # Chop of quotation marks
-        value = attr_parts[1]
-
-        if attr_parts[1].find('"') == 0:
-            value = attr_parts[1][1:]
-
-        if value.find('"') > -1:
-            value = value[0:-1]
-
         validate_header_attr(value, name=key)
-        value = unescape_header_attr(value)
+        if key in attributes:
+            raise BadHeaderValue('Duplicate key')
         attributes[key] = value
 
+    parsed_header = HAWK_HEADER_RE.sub(replace_attribute, attributes_full)
+    if parsed_header != '':
+        raise BadHeaderValue("Couldn't parse Hawk header")
```

Comment:
Yes, it's tempting to make the exception messages very obscure but I think ultimately the confusion it causes outweighs the safety. If anyone is exposing `mohawk` exception messages then they will already be compromising security; front-end interfaces like [hawkrest](https://github.com/kumar303/hawkrest/) have [comments](https://github.com/kumar303/hawkrest/blob/master/hawkrest/__init__.py#L82-L84) warning about this. I think `mohawk` is low level enough that it should be explicit about its failures.

Thanks for bringing this up though. I just made this more clear in the docs: https://github.com/kumar303/mohawk/pull/39